packages/amplify-graphql-model-transformer/src/resources/dynamo-model-resource-generator.ts

```diff
@@ -233,64 +233,70 @@ export class DynamoModelResourceGenerator extends ModelResourceGenerator {
   /**
    * createIAMRole
    */
-  createIAMRole = (context: TransformerContextProvider, def: ObjectTypeDefinitionNode, scope: Construct, tableName: string): iam.Role => {
+  createIAMRole = (context: TransformerContextProvider, def: ObjectTypeDefinitionNode, scope: Construct, tableName: string): iam.IRole => {
     const roleName = context.resourceHelper.generateIAMRoleName(ModelResourceIDs.ModelTableIAMRoleID(def!.name.value));
+    const amplifyDataStoreTableName = context.resourceHelper.generateTableName(SyncResourceIDs.syncTableName);
     const role = new iam.Role(scope, ModelResourceIDs.ModelTableIAMRoleID(def!.name.value), {
       roleName,
       assumedBy: new iam.ServicePrincipal('appsync.amazonaws.com'),
+      // Use an inline policy here to prevent unnecessary policy resources from being generated
+      // and slowing down deployments.
+      inlinePolicies: {
+        DynamoDBAccess: new iam.PolicyDocument({
+          statements: [
+            new iam.PolicyStatement({
+              effect: iam.Effect.ALLOW,
+              actions: [
+                'dynamodb:BatchGetItem',
+                'dynamodb:BatchWriteItem',
+                'dynamodb:PutItem',
+                'dynamodb:DeleteItem',
```

Comment:
Since we're customizing/overriding the CDK generated policy, do we have to exactly match it? Like if our state machine isn't performing any `DeleteItem`, can we skip it?

Reply:
This is actually the policy inlined to the role we generate for every model that the AppSync service principal assumes to operate on the model's data source (DynamoDB table).

Reply:
The policy statement for the `onEvent` / `isComplete` lambdas can be found [here](https://github.com/aws-amplify/amplify-category-api/pull/2490/files/92d96143e9b879da06817a9967fcef06db28cdc1#diff-2f1007f7f1c64b1c31c5f33e3efef71fa74eb48019015e14154109e08faae65eR57-R77) and the actions are unchanged.

Reply:
IIUC, this is the policy for the AppSync data source to connect to the underlying model table, not for the state machine. If I don't understand correctly, then I need to re-read this PR! :)

Reply:
Your understanding is correct 😄